File: xbox/src/main/java/com/microsoft/xal/browser/BrowserSelector.java
/*
 * Copyright (C) 2018-2022 Тимашков Иван
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.microsoft.xal.browser;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.pm.Signature;
import android.net.Uri;
import android.util.Base64;

import androidx.annotation.NonNull;
import androidx.browser.customtabs.CustomTabsService;
import androidx.core.os.EnvironmentCompat;

import com.microsoft.xal.logging.XalLogger;

import org.jetbrains.annotations.Contract;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;

/**
 * 13.08.2022
 *
 * @author Тимашков Иван
 * @author https://github.com/TimScriptov
 */
public class BrowserSelector {
    private static final Map<String, String> customTabsAllowedBrowsers;

    static {
        Map<String, String> hashMap = new HashMap<>();
        customTabsAllowedBrowsers = hashMap;
        hashMap.put("com.android.chrome", "OJGKRT0HGZNU+LGa8F7GViztV4g=");
        hashMap.put("org.mozilla.firefox", "kg9Idqale0pqL0zK9l99Kc4m/yw=");
        hashMap.put("com.microsoft.emmx", "P2QOJ59jvOpxCCrn6MfvotoBTK0=");
        hashMap.put("com.sec.android.app.sbrowser", "nKUXDzgZGd/gRG/NqxixmhQ7MWM=");
    }

    @NonNull
    public static BrowserSelectionResult selectBrowser(Context context, boolean z) {
        String str;
        XalLogger xalLogger = new XalLogger("BrowserSelector");
        try {
            BrowserSelectionResult.BrowserInfo userDefaultBrowserInfo = userDefaultBrowserInfo(context, xalLogger);
            boolean z2 = false;
            if (z) {
                str = "inProcRequested";
            } else if (browserInfoImpliesNoUserDefault(userDefaultBrowserInfo)) {
                str = "noDefault";
            } else {
                String str2 = userDefaultBrowserInfo.packageName;
                if (!browserSupportsCustomTabs(context, str2)) {
                    xalLogger.Important("selectBrowser() Default browser does not support custom tabs.");
                    str = "CTNotSupported";
                } else if (!browserAllowedForCustomTabs(context, xalLogger, str2)) {
                    xalLogger.Important("selectBrowser() Default browser supports custom tabs, but is not allowed.");
                    str = "CTSupportedButNotAllowed";
                } else {
                    xalLogger.Important("selectBrowser() Default browser supports custom tabs and is allowed.");
                    str = "CTSupportedAndAllowed";
                    z2 = true;
                }
            }
            BrowserSelectionResult browserSelectionResult = new BrowserSelectionResult(userDefaultBrowserInfo, str, z2);
            xalLogger.close();
            return browserSelectionResult;
        } catch (Throwable th) {
            try {
                throw th;
            } catch (Throwable th2) {
                if (th != null) {
                    try {
                        xalLogger.close();
                    } catch (Throwable th3) {
                        th.addSuppressed(th3);
                    }
                } else {
                    xalLogger.close();
                }
                throw th2;
            }
        }
    }

    @NonNull
    @Contract("_, _ -> new")
    private static BrowserSelectionResult.BrowserInfo userDefaultBrowserInfo(@NonNull Context context, XalLogger xalLogger) {
        String str;
        ResolveInfo resolveActivity = context.getPackageManager().resolveActivity(new Intent("android.intent.action.VIEW", Uri.parse("https://microsoft.com")), PackageManager.MATCH_DEFAULT_ONLY);
        String str2 = resolveActivity == null ? null : resolveActivity.activityInfo.packageName;
        if (str2 == null) {
            xalLogger.Important("userDefaultBrowserInfo() No default browser resolved.");
            return new BrowserSelectionResult.BrowserInfo("none", 0, "none");
        } else if (str2.equals("android")) {
            xalLogger.Important("userDefaultBrowserInfo() System resolved as default browser.");
            return new BrowserSelectionResult.BrowserInfo("android", 0, "none");
        } else {
            int i = -1;
            try {
                PackageInfo packageInfo = context.getPackageManager().getPackageInfo(str2, 0);
                i = packageInfo.versionCode;
                str = packageInfo.versionName;
            } catch (PackageManager.NameNotFoundException e) {
                xalLogger.Error("userDefaultBrowserInfo() Error in getPackageInfo(): " + e);
                str = EnvironmentCompat.MEDIA_UNKNOWN;
            }
            xalLogger.Important("userDefaultBrowserInfo() Found " + str2 + " as user's default browser.");
            return new BrowserSelectionResult.BrowserInfo(str2, i, str);
        }
    }

    @Contract(pure = true)
    private static boolean browserInfoImpliesNoUserDefault(@NonNull BrowserSelectionResult.BrowserInfo browserInfo) {
        return browserInfo.versionCode == 0 && browserInfo.versionName.equals("none");
    }

    @SuppressLint("PackageManagerGetSignatures")
    private static boolean browserAllowedForCustomTabs(Context context, XalLogger xalLogger, String str) {
        PackageInfo packageInfo;
        String str2 = customTabsAllowedBrowsers.get(str);
        if (str2 == null) {
            return false;
        }
        try {
            packageInfo = context.getPackageManager().getPackageInfo(str, PackageManager.GET_SIGNATURES);
            if (packageInfo == null) {
                xalLogger.Important("No package info found for package: " + str);
                return false;
            }
            for (Signature signature : packageInfo.signatures) {
                if (hashFromSignature(signature).equals(str2)) {
                    return true;
                }
            }
        } catch (PackageManager.NameNotFoundException e) {
            xalLogger.Error("browserAllowedForCustomTabs() Error in getPackageInfo(): " + e);
        } catch (NoSuchAlgorithmException e2) {
            xalLogger.Error("browserAllowedForCustomTabs() Error in hashFromSignature(): " + e2);
        }
        return false;
    }

    @SuppressLint("QueryPermissionsNeeded")
    private static boolean browserSupportsCustomTabs(@NonNull Context context, String str) {
        for (ResolveInfo resolveInfo : context.getPackageManager().queryIntentServices(new Intent(CustomTabsService.ACTION_CUSTOM_TABS_CONNECTION), 0)) {
            if (resolveInfo.serviceInfo.packageName.equals(str)) {
                return true;
            }
        }
        return false;
    }

    private static String hashFromSignature(@NonNull Signature signature) throws NoSuchAlgorithmException {
        MessageDigest messageDigest = MessageDigest.getInstance("SHA");
        messageDigest.update(signature.toByteArray());
        return Base64.encodeToString(messageDigest.digest(), 2);
    }
}

File: xbox/src/main/java/com/microsoft/xal/browser/BrowserSelector.kt
/*
 * Copyright (C) 2018-2022 Тимашков Иван
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.microsoft.xal.browser

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.content.pm.Signature
import android.net.Uri
import android.util.Base64
import android.util.Log
import androidx.browser.customtabs.CustomTabsService
import androidx.core.os.EnvironmentCompat
import com.microsoft.xal.browser.BrowserSelectionResult.BrowserInfo
import org.jetbrains.annotations.Contract
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

/**
 * 13.08.2022
 *
 * @author Тимашков Иван
 * @author https://github.com/TimScriptov
 */
object BrowserSelector {
    private val customTabsAllowedBrowsers = HashMap<String, String>()

    init {
        customTabsAllowedBrowsers.apply {
            put("com.android.chrome", "OJGKRT0HGZNU+LGa8F7GViztV4g=")
            put("org.mozilla.firefox", "kg9Idqale0pqL0zK9l99Kc4m/yw=")
            put("com.microsoft.emmx", "P2QOJ59jvOpxCCrn6MfvotoBTK0=")
            put("com.sec.android.app.sbrowser", "nKUXDzgZGd/gRG/NqxixmhQ7MWM=")
        }
    }

    @JvmStatic
    fun selectBrowser(context: Context, z: Boolean): BrowserSelectionResult {
        val notes: String
        val userDefaultBrowserInfo = userDefaultBrowserInfo(context)
        var useCustomTabs = false
        if (z) {
            notes = "inProcRequested"
        } else if (browserInfoImpliesNoUserDefault(userDefaultBrowserInfo)) {
            notes = "noDefault"
        } else {
            val packageName = userDefaultBrowserInfo.packageName
            if (!browserSupportsCustomTabs(context, packageName)) {
                Log.e("ModdedPE", "selectBrowser() Default browser does not support custom tabs.")
                notes = "CTNotSupported"
            } else if (!browserAllowedForCustomTabs(context, packageName)) {
                Log.e(
                    "ModdedPE",
                    "selectBrowser() Default browser supports custom tabs, but is not allowed."
                )
                notes = "CTSupportedButNotAllowed"
            } else {
                Log.e(
                    "ModdedPE",
                    "selectBrowser() Default browser supports custom tabs and is allowed."
                )
                notes = "CTSupportedAndAllowed"
                useCustomTabs = true
            }
        }
        return BrowserSelectionResult(userDefaultBrowserInfo, notes, useCustomTabs)
    }

    private fun userDefaultBrowserInfo(context: Context): BrowserInfo {
        var versionName: String
        val resolveActivity = context.packageManager.resolveActivity(
            Intent(
                "android.intent.action.VIEW",
                Uri.parse("https://microsoft.com")
            ), PackageManager.MATCH_DEFAULT_ONLY
        )
        return when (val packageName = resolveActivity?.activityInfo?.packageName) {
            null -> {
                Log.e("ModdedPE", "userDefaultBrowserInfo() No default browser resolved.")
                BrowserInfo("none", 0, "none")
            }
            "android" -> {
                Log.e("ModdedPE", "userDefaultBrowserInfo() System resolved as default browser.")
                BrowserInfo("android", 0, "none")
            }
            else -> {
                var versionCode = -1
                try {
                    val packageInfo = context.packageManager.getPackageInfo(packageName, 0)
                    versionCode = packageInfo.versionCode
                    versionName = packageInfo.versionName
                } catch (e: PackageManager.NameNotFoundException) {
                    Log.e("ModdedPE", "userDefaultBrowserInfo() Error in getPackageInfo(): $e")
                    versionName = EnvironmentCompat.MEDIA_UNKNOWN
                }
                Log.e(
                    "ModdedPE",
                    "userDefaultBrowserInfo() Found $packageName as user's default browser."
                )
                BrowserInfo(packageName, versionCode, versionName)
            }
        }
    }

    @Contract(pure = true)
    private fun browserInfoImpliesNoUserDefault(browserInfo: BrowserInfo): Boolean {
        return browserInfo.versionCode == 0 && browserInfo.versionName == "none"
    }

    @SuppressLint("PackageManagerGetSignatures")
    private fun browserAllowedForCustomTabs(context: Context, packageName: String): Boolean {
        val packageInfo: PackageInfo?
        val signatureBrowser = customTabsAllowedBrowsers[packageName] ?: return false
        try {
            packageInfo =
                context.packageManager.getPackageInfo(packageName, PackageManager.GET_SIGNATURES)
            if (packageInfo == null) {
                Log.e("ModdedPE", "No package info found for package: $packageName")
                return false
            }
            for (signature in packageInfo.signatures) {
                if (hashFromSignature(signature) == signatureBrowser) {
                    return true
                }
            }
        } catch (e: PackageManager.NameNotFoundException) {
            Log.e("ModdedPE", "browserAllowedForCustomTabs() Error in getPackageInfo(): $e")
        } catch (e: NoSuchAlgorithmException) {
            Log.e("ModdedPE", "browserAllowedForCustomTabs() Error in hashFromSignature(): $e")
        }
        return false
    }

    @SuppressLint("QueryPermissionsNeeded")
    private fun browserSupportsCustomTabs(context: Context, packageName: String): Boolean {
        for (resolveInfo in context.packageManager.queryIntentServices(
            Intent(CustomTabsService.ACTION_CUSTOM_TABS_CONNECTION),
            0
        )) {
            if (resolveInfo.serviceInfo.packageName == packageName) {
                return true
            }
        }
        return false
    }

    @Throws(NoSuchAlgorithmException::class)
    private fun hashFromSignature(signature: Signature): String {
        return Base64.encodeToString(MessageDigest.getInstance("SHA").apply {
            update(signature.toByteArray())
        }.digest(), 2)
    }
}
